README
# BEGIN BPS TAGGED BLOCK {{{
# COPYRIGHT:
# 
# This software is Copyright (c) 2003-2006 Best Practical Solutions, LLC
#                                          <clkao@bestpractical.com>
# 
# (Except where explicitly superseded by other copyright notices)
# 
# 
# LICENSE:
# 
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of either:
# 
#   a) Version 2 of the GNU General Public License.  You should have
#      received a copy of the GNU General Public License along with this
#      program.  If not, write to the Free Software Foundation, Inc., 51
#      Franklin Street, Fifth Floor, Boston, MA 02110-1301 or visit
#      their web page on the internet at
#      http://www.gnu.org/copyleft/gpl.html.
# 
#   b) Version 1 of Perl's "Artistic License".  You should have received
#      a copy of the Artistic License with this package, in the file
#      named "ARTISTIC".  The license is also available at
#      http://opensource.org/licenses/artistic-license.php.
# 
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
# 
# CONTRIBUTION SUBMISSION POLICY:
# 
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of the
# GNU General Public License and is only of importance to you if you
# choose to contribute your changes and enhancements to the community
# by submitting them to Best Practical Solutions, LLC.)
# 
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with SVK,
# to Best Practical Solutions, LLC, you confirm that you are the
# copyright holder for those contributions and you grant Best Practical
# Solutions, LLC a nonexclusive, worldwide, irrevocable, royalty-free,
# perpetual, license to use, copy, create derivative works based on
# those contributions, and sublicense and distribute those contributions
# and any derivatives thereof.
# 
# END BPS TAGGED BLOCK }}}
This is the README file for SVK, a decentralized version control
system.

The latest version could be found at

    http://svn.bestpractical.com/svk/trunk

You need Module::Install if you are not installing from release tarballs.

* About

While Subversion aims to take over the CVS user base, SVK attempts
to take over the others -- including people who are already using
other SCM systems, and people who had not started using SCM.

SVK is written in Perl, using Subversion's underlying filesystem
but provides additional features:

 * Offline operations like "checkin", "log", "merge".
 * Distributed branches.
 * Lightweight checkout copy management (no ".svn" directories).
 * Advanced merge algorithms, like "star-merge" and "cherry picking".

For more information, please visit the web site at http://svk.bestpractical.com/.

* Installation

To use SVK, you need Subversion version 1.0.4 or higher, compiled with
Perl bindings support.

SVK uses the standard perl module install process:

% perl Makefile.PL
% make
# make install

* Usage

You may want to start with the SVK tutorial, available on the web site.


CHANGES
[Changes for 2.2.1 - 10 Mar, 2010]

  Subversion 1.6.x compatiblity fixes.

[Changes for 2.2.1 - 22 Oct, 2008]

  Fix the incorrect version information in META.yml

  Mirror subsystem
    * Fix memory leak for bootstrapping mirrors.

[Changes for 2.2.0 - 17 Sep, 2008]

  New features
    * Branch support as 'svk branch' subcommand
    * SVKBATCHMODE environment variable

  Dependency
    * Require Time::Progress

  Mirror subsystem
    * New bootstrap option to initialize a mirror from an svn dump
      file, which can be an URL stored on the svk:dump-url property on
      the server
    * When committing or merging to a mirror, avoid roundtripping
      the latest revision (requires Subversion 1.5)
    * Fix a bug when mirroring a copy revision with unnormalized
      revision outside mirror anchor and newer th an anchor head
    * Fix URI escape issues for pushing changes to mirrors
    * Support mirroring a source that was renamed from somewhere else
    * Fix svk sync where a copy is made from revisions condensed with
      snapshot made by previous sync --skipto

  Merge subsystem
    * Bugfixes for opened file handle passed to editor under win32,
      causing file to be opened as read-only
    * Fix a problem that merge-ticket might be reverted on mirror
      in some cases
    * Fix a bug when merging a copy with modification to the copied
      source, it might cause checksum error
    * Addition of equality check for merge tickets
    * Better conflict handling that remembers the type of conflict,
      handling of data from the merge editor

  Tests

    * Test coverage for offlining existing branches (both offline and
      online)
    * TODO tests for renaming of an intermediate branch tracking
    * Test coverage for pulling into a branch from a local branch
    * Test coverage for checking out to a path from svk br --checkout 
    * Test coverage for modifying property changes (including conflicts)
    * Fixed tests for baseless smerge
    * Test coverage for switching a branch and proper handling of
      merge tickets

  General 
    * General Documentation cleanup
    * Documentation for SVK::Info properly matches expected output
    * Conversion to use SVK::Log (and Log::Log4perl) for output
    * Add an option of '0' to select "No project" if there is not a
      project to use while mirroring a new URI
    * svk will not set property status as modified when a path is added or 
      replaced
    * svk now stores known conflict types (provided by merge editor) in 
      .conflict property
    * General usability for working with files no longer exists in HEAD
     
  svk commit
    * A property may have a conflict when the node is ok, so combine
      them in the target list

  svk merge
    * Don't write to the log if it is a check run, regardless of -l option
  
  svk resolved
    * svk resolved will resolve all conflicts in a checkout path

  svk update
    * Add custom svk:merge property resolver handler to handle merging
      of merge tickets properly

  Internal
    * svk will not clobber upstream Log::Log4perl configurations



Makefile.PL
# BEGIN BPS TAGGED BLOCK {{{
# COPYRIGHT:
# 
# This software is Copyright (c) 2003-2008 Best Practical Solutions, LLC
#                                          <clkao@bestpractical.com>
# 
# (Except where explicitly superseded by other copyright notices)
# 
# 
# LICENSE:
# 
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of either:
# 
#   a) Version 2 of the GNU General Public License.  You should have
#      received a copy of the GNU General Public License along with this
#      program.  If not, write to the Free Software Foundation, Inc., 51
#      Franklin Street, Fifth Floor, Boston, MA 02110-1301 or visit
#      their web page on the internet at
#      http://www.gnu.org/copyleft/gpl.html.
# 
#   b) Version 1 of Perl's "Artistic License".  You should have received
#      a copy of the Artistic License with this package, in the file
#      named "ARTISTIC".  The license is also available at
#      http://opensource.org/licenses/artistic-license.php.
# 
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
# 
# CONTRIBUTION SUBMISSION POLICY:
# 
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of the
# GNU General Public License and is only of importance to you if you
# choose to contribute your changes and enhancements to the community
# by submitting them to Best Practical Solutions, LLC.)
# 
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with SVK,
# to Best Practical Solutions, LLC, you confirm that you are the
# copyright holder for those contributions and you grant Best Practical
# Solutions, LLC a nonexclusive, worldwide, irrevocable, royalty-free,
# perpetual, license to use, copy, create derivative works based on
# those contributions, and sublicense and distribute those contributions
# and any derivatives thereof.
# 
# END BPS TAGGED BLOCK }}}
#!/usr/bin/perl
my $required_svn = '1.3.0';
my $required_svm = '0.71';

use strict;
use inc::Module::Install;

name		('SVK');
author		('Chia-liang Kao <clkao@clkao.org>');
abstract	('A decentralized version control system');
license		('perl');
version_from	('lib/SVK/Version.pm');
install_script	('bin/svk');
clean_files	('t/checkout', 't/apache_svn');
requires(
    # 'SVN::Core'                => $required_svn,
    'version'                  => '0.68',
    'Algorithm::Annotate'      => '0',
    'Algorithm::Diff'          => '1.1901',
    'YAML::Syck'               => '0.60',
    'Data::Hierarchy'          => '0.30',
    'PerlIO::via::dynamic'     => '0.11',
    'PerlIO::via::symlink'     => '0.02',
    'IO::Digest'               => '0',
    'SVN::Simple::Edit'        => '0.27',
    'URI'                      => '0',
    'PerlIO::eol'              => '0.13',
    'Class::Autouse'           => '1.15',
    'App::CLI'                 => '0',
    'List::MoreUtils'          => '0',
    'Class::Accessor::Fast'    => '0',
    'Class::Data::Inheritable' => '0',
    'Path::Class'              => '0.16',
    'UNIVERSAL::require'       => '0',
    'Term::ReadKey'            => '0',
    'Time::HiRes'              => '0',       # in core since 5.7.3
    'File::Temp'               => '0.17',
    'Encode'                   => '2.10',    # in core since 5.8.7
    'Getopt::Long'             => '2.35',    # in core since 5.8.8
    'Pod::Escapes'             => '0',       # in core since 5.9.3
    'Pod::Simple'              => '0',       # in core since 5.9.3
    'File::Spec'               => '3.19',    # in core since 5.9.3
    'Time::Progress'           => '0',
    'Git::PurePerl'            => '0.38',
);


requires(
    'Date::Format'             => '',
) if ($^O eq 'MSWin32');
build_requires(
    'Test::More'               => '0.42',
);
features(
    'Bootstrap support' => [
    -default => 1,
        'PerlIO::via::Bzip2'       => '0',
        'PerlIO::gzip'             => '0',
        'SVN::Dump'                => '0.04',
        'LWP::UserAgent'           => '0',
    ],
    'Localized messages' => [
	-default => 1,
        'Locale::Maketext::Lexicon' => '0.62',
        'Locale::Maketext::Simple'  => '0.16',
    ],
    'Message display through a pager' => [
        -default => 1,
        'IO::Pager'             => '0',
    ],
    'Log4perl support' => [
        -default => 0,
        'Log::Log4perl'         => '0',
    ]
);


if( eval{ require SVN::Mirror } ) {
    requires( 'SVN::Mirror' => $required_svm );
} else {
    features( 'SVN::Mirror support' => [
               -default => 1,
               'SVN::Mirror'    => $required_svm,
           ],
	    );
}

features(
    'Patch management' => [
	-default => 1,
	'Compress::Zlib'    => '0',
	'FreezeThaw'        => '0',
    ],
);

# Include subdirectory tests too.
tests("t/*.t t/*/*.t");

include('ExtUtils::AutoInstall');
auto_install();

auto_provides();
WriteAll( sign => 0 );

$::VERSION = $required_svn;
die << "." unless eval {require SVN::Core; SVN::Core->VERSION >= main->VERSION };
===> Warning: SVN::Core missing or outdated.
Please manually install SVN::Core by compiling Subversion
(version $required_svn or above) with SWIG/Perl bindings,
and try this installation process again.
.
